tasks/files-to-granules/tests/test-index.js

```diff
@@ -52,3 +56,67 @@ test('files-to-granules transforms files array to granules object', async (t) =>
   await validateOutput(t, output);
   t.deepEqual(output, expectedOutput);
 });
+
+test('files-to-granules matches granules using producerGranuleId if configured', async (t) => {
+  const event = t.context.payload;
+
+  // Make the payload granuleId unique
+  const granuleDateString = '2017201090724';
+  const granuleIdReplacement = cryptoRandomString({ length: 13, type: 'numeric' });
+  event.config.inputGranules[0].granuleId = event.config.inputGranules[0].granuleId
+    .replace(granuleDateString, granuleIdReplacement);
+
+  event.config.matchFilesWithProducerGranuleId = true;
+
+  await validateConfig(t, event.config);
+  await validateInput(t, event.input);
+  const expectedOutput = t.context.output;
+  // Make the expected output granuleId unique
+  expectedOutput.granules[0].granuleId = expectedOutput.granules[0].granuleId
+    .replace(granuleDateString, granuleIdReplacement);
+  const output = await filesToGranules(event);
+  await validateOutput(t, output);
+
+  t.deepEqual(output, expectedOutput);
+});
+
+test('files-to-granules throws error if configured to use ID that does not exist', async (t) => {
+  const event = t.context.payload;
+
+  // Make the payload granuleId unique
+  const granuleDateString = '2017201090724';
+  const granuleIdReplacement = cryptoRandomString({ length: 13, type: 'numeric' });
+  event.config.inputGranules[0].granuleId = event.config.inputGranules[0].granuleId
+    .replace(granuleDateString, granuleIdReplacement);
+
+  event.config.inputGranules[0].producerGranuleId = '';
+  event.config.matchFilesWithProducerGranuleId = true;
+
+  await validateConfig(t, event.config);
+  await validateInput(t, event.input);
+
+  await t.throwsAsync(
+    filesToGranules(event),
+    { instanceOf: UnmetRequirementsError }
+  );
+});
+
+test('files-to-granules throws error if configured to use uniquified granuleId for matching files', async (t) => {
+  const event = t.context.payload;
+
+  // Make the payload granuleId unique
+  const granuleDateString = '2017201090724';
+  const granuleIdReplacement = cryptoRandomString({ length: 13, type: 'numeric' });
+  event.config.inputGranules[0].granuleId = event.config.inputGranules[0].granuleId
+    .replace(granuleDateString, granuleIdReplacement);
+
+  event.config.matchFilesWithProducerGranuleId = true;
+
+  await validateConfig(t, event.config);
+  await validateInput(t, event.input);
+
+  await t.throwsAsync(
+    filesToGranules(event, { fileObjectFromS3URI: () => Promise.reject() }),
+    { instanceOf: Error }
```

Comment:
You could argue this isn't a great test as the other error path is also a child of `Error`. The object here was coverage and the trow we're testing in this assertion is the 'else' fallback in case something very unexpected happens so I wasn't too fussed about it. Happy to adjust but didn't feel like it was worth e.g. changing the error type thrown.